src/UserSpaceInstrumentation/InstrumentProcess.cpp

```diff
@@ -0,0 +1,441 @@
+  if (!trampolines_for_modules_.contains(address_range)) {
+    trampolines_for_modules_.emplace(address_range, TrampolineMemoryChunks());
+  }
+  auto it = trampolines_for_modules_.find(address_range);
+  TrampolineMemoryChunks& trampoline_memory_chunks = it->second;
+  if (trampoline_memory_chunks.empty() ||
+      trampoline_memory_chunks.back().first_available == kTrampolinesPerChunk) {
+    OUTCOME_TRY(auto&& trampoline_memory,
+                AllocateMemoryForTrampolines(pid_, address_range,
+                                             kTrampolinesPerChunk * GetMaxTrampolineSize()));
+
+    trampoline_memory_chunks.emplace_back(std::move(trampoline_memory), 0);
+  }
+  const uint64_t result = trampoline_memory_chunks.back().memory->GetAddress() +
+                          trampoline_memory_chunks.back().first_available * GetMaxTrampolineSize();
+  trampoline_memory_chunks.back().first_available++;
+  return result;
+}
+
+ErrorMessageOr<void> InstrumentedProcess::ReleaseMostRecentlyAllocatedTrampolineMemory(
+    AddressRange address_range) {
+  if (!trampolines_for_modules_.contains(address_range)) {
+    return ErrorMessage("Tried to release trampoline memory for a non existent address range");
+  }
+  trampolines_for_modules_.find(address_range)->second.back().first_available--;
+  return outcome::success();
+}
+
+ErrorMessageOr<std::vector<ModuleInfo>> InstrumentedProcess::ModulesFromModulePath(
+    std::string path) {
+  if (!modules_from_path_.contains(path)) {
+    std::vector<ModuleInfo> result;
+    OUTCOME_TRY(auto&& modules, orbit_object_utils::ReadModules(pid_));
+    for (const ModuleInfo& module : modules) {
+      if (module.file_path() == path) {
+        result.push_back(module);
+      }
+    }
+    if (result.empty()) {
+      return ErrorMessage(absl::StrFormat("Unable to find module for path %s", path));
+    }
+    modules_from_path_.emplace(path, result);
+  }
+  return modules_from_path_.find(path)->second;
+}
+
+ErrorMessageOr<void> InstrumentedProcess::EnsureTrampolinesWritable() {
+  for (auto& trampoline_for_module : trampolines_for_modules_) {
+    for (auto& memory_chunk : trampoline_for_module.second) {
+      OUTCOME_TRY(memory_chunk.memory->EnsureMemoryWritable());
+    }
+  }
+  return outcome::success();
+}
+
+ErrorMessageOr<void> InstrumentedProcess::EnsureTrampolinesExecutable() {
+  for (auto& trampoline_for_module : trampolines_for_modules_) {
+    for (auto& memory_chunk : trampoline_for_module.second) {
+      OUTCOME_TRY(memory_chunk.memory->EnsureMemoryExecutable());
+    }
+  }
+  return outcome::success();
+}
+
+std::unique_ptr<UserSpaceInstrumentation> UserSpaceInstrumentation::Create() {
+  static bool first_call = true;
+  FAIL_IF(!first_call, "UserSpaceInstrumentation should be globally unique.");
+  first_call = false;
+  return std::unique_ptr<UserSpaceInstrumentation>(new UserSpaceInstrumentation());
```

Comment:
If this is about global uniqueness I guess it should be thread safe, but two threads could both find `first_call` true.

Reply:
I don't think that this will ever be an issue but I through in a mutex for good measure.